src/python/xml_parser-py.c

```diff
@@ -69,14 +69,10 @@ c_newpkgcb(cr_Package **pkg,
         *pkg = NULL;
     } else {
         *pkg = Package_FromPyObject(result);
-        if (data->py_pkgcb != Py_None) {
-            // Store reference to the python pkg (result) only if we will need it later
-            PyObject *keyFromPtr = PyLong_FromVoidPtr(*pkg);
-            PyDict_SetItem(data->py_pkgs, keyFromPtr, result);
-            Py_DECREF(keyFromPtr);
-        }
```

Comment:
In the event that there is no `py_pkgcb`, the only reference to the underlying `cr_Package` is held via `result`, and that refcount goes to 0 on line 80 below, and it gets freed.  But since this is `newpkgcb` this the package isn't finished being parsed yet so the parsing code is writing to a dead object.

The code used to wait until the next iteration (once the package has finished parsing) before decrementing the refcount, so it was safe: https://github.com/rpm-software-management/createrepo_c/commit/670006b9f37d8298fe2ec1d6cfc5cdf175a3128b#diff-0a3304524b67e6c30420cf0bd3f66c5937250b1e6837414a0d9d0ea5bc87ab63L52

Since the reference needs to be stored no matter what, I just picked the approach of simplifying the codepaths and using one method, but let me know if you have something else in mind.

Reply:
I see, that is a good catch!
I would prefer a different solution tough.

Because if we do this then anytime we use the API without `py_pkgcb` we store all the packages in memory throughout the run because `data->py_pkgs` is cleared only in the `py_pkgcb` callback (not set here) or at the very end so at that point we might as well use the old API.
It would also be pointless work to parse the pkg in such a case since we know the user doesn't have any reference to it and has no way to get it either since there is no `py_pkgcb` (at least for this version of the API).

Right now it would make most sense to me to check if the package gets freed by the `Py_DECREF` and if so set `*pkg` to `NULL` to skip the package.

Perhaps the end of the function could look like this (not sure if there is some better way to do this)?
```
     if (result->ob_refcnt == 1) {
          *pkg = NULL;
     }
     Py_DECREF(result);

     return CR_CB_RET_OK;
```

Reply:
It feels a little unsatisfying but I'm not sure of a better way.  I've made the change though and Valgrind passes (apart from the one issue I mentioned already below)